### ISSUE TITLE
Support package-level registries

### ISSUE THREAD

**Issue by kdex** [jspm/jspm-cli#1691]
Right now, if you have a bower dependency, adding `jspm-bower-endpoint` to your `package.json` won't cause the bower registry to be used. The only way to use it is to install it globally, which makes packages with bower dependencies particularly hard to ship.

This, of course, doesn't only affect bower, but really any registry that needs a registry/endpoint.

Would it be feasible for jspm to use locally installed registries in `node_modules`?

**Comment by guybedford** [comment]
Yes this should definitely be supported.

This would be done by altering the require at https://github.com/jspm/jspm-cli/blob/master/lib/registry.js#L68 to be a project-level require, based to the project folder itself.

The issue here is I don't know how to use `require('module')` in Node to generate a custom require that will work here.

I've tried:

```javascript
var Module = require('module');
function scopedRequire(name) {
  return Module._load(name, new Module(process.env.jspmConfigPath && path.dirname(process.env.jspmConfigPath) || path.resolve(process.cwd())), true);
};
```

to replace the require above, but this doesn't seem to work properly. A variation of this technique should work though.

**Comment by Zenorbi** [comment]
@guybedford I'm currently trying to achieve something similar in jspm. Can you please elaborate on what would the `scopedRequire` and how would the technique work? What I think would be a good solution is this:

1. Whenever a package needs a custom registry, the user of the package should npm install the required registry handler locally.
2. When the package uses custom registry, the `package.json` should inform JSPM about what is the name of the registry and what is the node module name of its handler.
3. JSPM uses the registry info from the `package.json` in a scoped manner, meaning multiple packages can use the same registry with different handlers.

**Reasons**

The 1. point ensures that the user of the package can use custom configs for the handler. (Like custom authentications, tokens, proxies.) JSPM also doesn't need to npm install in the middle of the dependency resolution. This is like using peer dependency in npm.

The 2. point ensures that if the package wants to use `bower:*`, the node module name of the handler can be resolved, jspm can inform the user as to what registry the it should install locally (or globally) and the package can use its custom registry name.

The 3. point ensures that the package doesn't need to use very long handlers to avoid collision.

**Example `package.json`**

```
{
  "jspm": {
    "name": "my-package",
    "version": "0.1.0",
    "registries": {
      "bower": {
        "handler": "jspm-bower-endpoint",
        "some-default-config": "foo"
      },
      "dependencies": {
        "some-dep": "bower:some-dependency@1.0.0"
      }
    }
  }
}
```

**Example workflow**

```
> npm install jspm --save-dev

> jspm install github:package-that-uses-bower

< Package named "github:package-that-uses-bower" needs a custom
< registry named "jspm-bower-endpoint". Please install the registry
< using `npm install jspm-bower-endpoint --save-dev`.

> npm install jspm-bower-endpoint --save-dev

> jspm install github:package-that-uses-bower

< Done 
```

**Comment by guybedford** [comment]
Implemented in https://github.com/jspm/jspm-cli/commit/0901db4ca93a07163bafe2ea062d8bc2e49a3f6b.

**Comment by guybedford** [comment]
Released in jspm 0.17.0-beta.15.
